### ISSUE TITLE
<input> minlength numeric attribute not being set on underlying element

### ISSUE THREAD

**Issue by spiffytech** [forgojs/forgo#31]
[Sandbox](https://codesandbox.io/s/forgo-minlength-missing-sonic)

When I render `<input minlength={8} />` Forgo doesn't set the `minlength` attribute on the underlying DOM element.

_I'm using `={8}` because TypeScript insists the value needs to be a number, but if I do a string anyway the attribute still doesn't get set._

**Comment by spiffytech** [comment]
It looks like this is because `<input>` elements don't have a `.minlength` property, and setting the property only affects attributes if that property is specifically supported. I.e., `el["x-whatever"] = "abc"` has no effect on the element's attributes, because `x-whatever` isn't an official, spec'd attribute, but `el["classList"] = "abc"` works because that's a recognized property.

So if we change `attachProps()` to look like this:

```ts
            // This optimization is copied from preact.
            else if (
              typeof value === "string" &&
              (key.startsWith("aria-") || key.startsWith("data-"))
            ) {
              (node as HTMLElement).setAttribute(key, value);
            } else if (key === "onblur") {
              (node as any)[key] = handlerDisabledOnNodeDelete(node, value);
            } else if (Reflect.has(node, key)) {
              (node as any)[key] = value;
            } else {
              (node as any).setAttribute(key, value);
            }
```

Then the input correctly gets its `maxlength` attribute set. This change also covers the `aria-` and `data-` special cases, so maybe those could be removed, too?

While the `maxlength` attribute is what I ran into, this revealse that there's a broader bug that Forgo won't set attributes like Alpine's `x-data` or petite-vue's `v-scope`. This change fixes that, too.

Do you have a handy benchmark to see the impact `Reflect.has()` has on performance? `.hasOwnProperty()` doesn't do the trick, since it looks like built-in properties can come from the prototype chain, which `.hasOwnProperty()` won't detect.

**Comment by jeswin** [comment]
Wow, thanks for the details and the code sample. I'll look at it tomorrow; and review your fix as well.

**Comment by jeswin** [comment]
Sorry I got stuck in breadwinning work yesterday. :)

1. Reflect.has() seems to have good support - https://caniuse.com/mdn-javascript_builtins_reflect
2. My naive unscientific performance tests give a 4x advantage to `key in el` over Reflect.has()

So go with `key in el`?

**Comment by spiffytech** [comment]
`key in el` sounds good to me.

**Comment by jeswin** [comment]
Excellent bug report and solution. Fixed in v2.1.0.

**Comment by spiffytech** [comment]
Confirmed, thanks!
